src/common/src/cache.rs

```diff
@@ -769,65 +785,102 @@ impl<K: LruKey, T: LruValue> LruCache<K, T> {
+        self.key.take().unwrap()
+    }
 }
 
 impl<'a, K: LruKey + Clone + 'static, T: LruValue + 'static> Drop for CleanCacheGuard<'a, K, T> {
     fn drop(&mut self) {
-        if !self.success {
-            self.cache.clear_pending_request(&self.key, self.hash);
+        if let Some(key) = self.key.as_ref() {
+            self.cache.clear_pending_request(key, self.hash);
+        }
+    }
+}
+
+/// `lookup_with_request_dedup` can directly return `Result<CacheableEntry<K, T>, E>`, but if we do
+/// not want to wait when cache hit does not happen, we can call `lookup_with_request_dedup_noawait`
+/// which will return `BoxFuture<'static, Result<CacheableEntry<K, T>, E>>` when cache hit does not
+/// happen.
+pub enum LookupResponse<K: LruKey + Clone + 'static, T: LruValue + 'static, E> {
+    Cached(CacheableEntry<K, T>),
+    WaitPendingRequest(Receiver<CacheableEntry<K, T>>),
+    Miss(JoinHandle<Result<CacheableEntry<K, T>, E>>),
+}
+
+impl<K: LruKey + Clone + 'static, T: LruValue + 'static, E: From<RecvError>>
+    LookupResponse<K, T, E>
+{
+    pub async fn wait(self) -> Result<CacheableEntry<K, T>, E> {
+        match self {
+            LookupResponse::Cached(entry) => Ok(entry),
+            LookupResponse::WaitPendingRequest(receiver) => {
+                receiver.await.map_err(|recv_error| recv_error.into())
+            }
+            LookupResponse::Miss(join_handle) => join_handle.await.unwrap(),
         }
     }
 }
 
-/// Only implement `lookup_with_request_dedup` for static values, as they can be sent across tokio
-/// spawned futures.
+/// Only implement `lookup_with_request_dedup` and `lookup_with_request_dedup_noawait` for static
+/// values, as they can be sent across tokio spawned futures.
 impl<K: LruKey + Clone + 'static, T: LruValue + 'static> LruCache<K, T> {
-    pub async fn lookup_with_request_dedup<F, E, VC>(
+    pub fn lookup_with_request_dedup_noawait<F, E, VC>(
         self: &Arc<Self>,
         hash: u64,
         key: K,
         fetch_value: F,
-    ) -> Result<CacheableEntry<K, T>, E>
+    ) -> LookupResponse<K, T, E>
     where
         F: FnOnce() -> VC,
-        E: Error + Send + 'static,
+        E: Error + Send + 'static + From<RecvError>,
         VC: Future<Output = Result<(T, usize), E>> + Send + 'static,
     {
-        loop {
-            match self.lookup_for_request(hash, key.clone()) {
-                LookupResult::Cached(entry) => return Ok(entry),
-                LookupResult::WaitPendingRequest(recv) => {
-                    let _ = recv.await;
-                    continue;
-                }
-                LookupResult::Miss => {
-                    let this = self.clone();
-                    let fetch_value = fetch_value();
-                    let key2 = key.clone();
-                    let mut guard = CleanCacheGuard {
-                        cache: self,
-                        key,
+        match self.lookup_for_request(hash, key.clone()) {
```

Comment:
Our current code, via the loop, will trigger an immediate retry by other previously pending requests when the unique request that does the IO fails. 
 After this PR, seems that such retry not longer exists. I am not sure whether this impact our code. cc @hzxa21 @Little-Wallace

Reply:
> Our current code, via the loop, will trigger an immediate retry by other previously pending requests when the unique request that does the IO fails.
> After this PR, seems that such retry not longer exists. I am not sure whether this impact our code. cc @hzxa21 @Little-Wallace

@hzxa21 thought it acceptable

Reply:
Previously in the `LookupResut::Miss` branch, we anyway won't retry and just epoxse the error to the caller. Therefore, I think it is acceptable to maintain the same behavior in the `LookupResult::WaitPendingRequest` branch. Not sure whether there is any corner case that can cause deadlock. What do you think? @Little-Wallace

Reply:
> Previously in the `LookupResut::Miss` branch, we anyway won't retry and just epoxse the error to the caller. Therefore, I think it is acceptable to maintain the same behavior in the `LookupResult::WaitPendingRequest` branch. Not sure whether there is any corner case that can cause deadlock. What do you think? @Little-Wallace


I remember this loop because the Receiver of WaitPendingRequest did not return CacheableEntry (a previous implementation to avoid deadlock), so the cache needs to be queried again to get the result. So I think loop can be removed.

it is acceptable to maintain the same behavior that epoxse the error to the caller

Reply:
I add this loop because the IO thread may cancel the future and make all other thread failed. See code in https://github.com/risingwavelabs/risingwave/blob/4e7bb2ccce73e126f51ed3827c45a634b5743560/src/common/src/cache.rs#L824
If the previous thread is canceled by some join executor, all other thread would fail.

Reply:
> I add this loop because the IO thread may cancel the future and make all other thread failed. See code in
> 
> https://github.com/risingwavelabs/risingwave/blob/4e7bb2ccce73e126f51ed3827c45a634b5743560/src/common/src/cache.rs#L824
> 
> 
> If the previous thread is canceled by some join executor, all other thread would fail.

but when cache miss happens, we will `tokio::spawn` a task to fetch block, so I think that even the previous thread is canceled, the spawned task will still continue.